### PR TITLE
Replace EasyMock and PowerMock with Mockito | TimeOrderedCachingPersistentWindowStoreTest

### PR DESCRIPTION
**Pull request by ghost**
Related to KAFKA-14059 and KAFKA-14132

Link: 

* https://issues.apache.org/jira/browse/KAFKA-14133
* https://issues.apache.org/jira/browse/KAFKA-14132

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 1]
```diff
@@ -124,7 +129,8 @@ public static Collection<Object[]> data() {
     public void setUp() {
         baseKeySchema = new TimeFirstWindowKeySchema();
         bytesStore = new RocksDBTimeOrderedWindowSegmentedBytesStore("test", "metrics-scope", 100, SEGMENT_INTERVAL, hasIndex);
-        underlyingStore = new RocksDBTimeOrderedWindowStore(bytesStore, false, WINDOW_SIZE);
+        underlyingStore = spy(new RocksDBTimeOrderedWindowStore(bytesStore,
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Is this intentional?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 2]
```diff
@@ -138,51 +144,69 @@ public void setUp() {
...
     @Test
     public void shouldDelegateDeprecatedInit() {
-        final RocksDBTimeOrderedWindowStore inner = EasyMock.mock(RocksDBTimeOrderedWindowStore.class);
-        EasyMock.expect(inner.hasIndex()).andReturn(hasIndex);
-        EasyMock.replay(inner);
-        final TimeOrderedCachingWindowStore outer = new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE, SEGMENT_INTERVAL);
+        final RocksDBTimeOrderedWindowStore inner = mock(RocksDBTimeOrderedWindowStore.class);
+        when(inner.hasIndex()).thenReturn(hasIndex);
+
+        final TimeOrderedCachingWindowStore outer =
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: I think we do not need to spy on the outer, since here we are only checking that when `outer.init` is called, its `inner.init` is called too. Ditto for other tests

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 1 above)
Comment (ghost) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Yeah, this will help in `reset`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 2 above)
Comment (ghost) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Thanks! Updated.

**Comment by ghost** [comment]
I am not sure why one test case failed : https://github.com/apache/kafka/pull/12739/checks?check_run_id=8869541214 . I have not done any change in any other test file.

**Comment by ghost** [comment]
Not sure why these checks failed. These failures are not looking relevant. Please someone help.

**Comment by ghost** [comment]
Failing tests seem unrelated 

[
kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsToZonedDateTime, 
kafka.admin.MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful
]

**Comment by ghost** [comment]
Thanks @guozhangwang for the review. I have updated the PR, please have a look.

**Comment by ghost** [comment]
Ping for the PR review @C0urante  
Thanks!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 3]
```diff
@@ -1159,58 +1183,42 @@ public void shouldNotThrowInvalidBackwardRangeExceptionWithNegativeFromKey() {
...
     @Test
     public void shouldCloseCacheAfterErrorDuringStateStoreClose() {
         setUpCloseTests();
-        EasyMock.reset(cache);
+        reset(cache);
         cache.flush(CACHE_NAMESPACE);
         cache.close(CACHE_NAMESPACE);
-        EasyMock.replay(cache);
-        EasyMock.reset(underlyingStore);
-        underlyingStore.close();
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:1199]: please verify invocation for this

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 4]
```diff
@@ -90,8 +88,15 @@
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertThrows;
 import static org.junit.Assert.assertTrue;
-
-@RunWith(Parameterized.class)
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:94]: We are removing the ability to run the tests in a parameterized manner. Hence, this test will not for parameters defined in 
```
@Parameterized.Parameters(name = "{0}")
    public static Collection<Object[]> data() {
        return asList(new Object[][] {
            {true},
            {false}
        });
    }
```

I understand that you did this because we cannot have two `RunWith` annotations. You can alternatively use 
```
@Rule
    public MockitoRule mockito = MockitoJUnit.rule().strictness(Strictness.STRICT_STUBS);
```
instead of the `@RunWith(MockitoJUnitRunner.StrictStubs.class)`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 5]
```diff
@@ -138,51 +143,70 @@ public void setUp() {
 
     @After
     public void closeStore() {
-        cachingStore.close();
+        try {
+            cachingStore.close();
+        } catch (final RuntimeException runtimeException) {
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: This may hide cases when the exception is genuine.

Alternatively, create a function:
```
private void verifyAndTearDownCloseTests() {
        verify(underlyingStore).close();
        verify(cache).flush(CACHE_NAMESPACE);
        verify(cache).close(CACHE_NAMESPACE);

        // resets the mocks created in #setUpCloseTests(). It is necessary to
        // ensure that @After works correctly.
        reset(cache);
        reset(underlyingStore);
    }
```

And call it at the end of each close test.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 6]
```diff
@@ -1159,58 +1183,42 @@ public void shouldNotThrowInvalidBackwardRangeExceptionWithNegativeFromKey() {
...
     private void setUpCloseTests() {
-        underlyingStore = EasyMock.createNiceMock(RocksDBTimeOrderedWindowStore.class);
-        EasyMock.expect(underlyingStore.name()).andStubReturn("store-name");
-        EasyMock.expect(underlyingStore.isOpen()).andStubReturn(true);
-        EasyMock.replay(underlyingStore);
-        cachingStore = new TimeOrderedCachingWindowStore(underlyingStore, WINDOW_SIZE, SEGMENT_INTERVAL);
-        cache = EasyMock.createNiceMock(ThreadCache.class);
+        underlyingStore = mock(RocksDBTimeOrderedWindowStore.class);
+        when(underlyingStore.name()).thenReturn("store-name");
+        cachingStore = spy(new TimeOrderedCachingWindowStore(underlyingStore,
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: I wasn't able to understand this change. Why do we need to spy this?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 7]
```diff
@@ -1159,58 +1183,42 @@ public void shouldNotThrowInvalidBackwardRangeExceptionWithNegativeFromKey() {
...
         assertThrows(RuntimeException.class, cachingStore::close);
-        EasyMock.verify(cache, underlyingStore);
     }
 
     @Test
     public void shouldCloseCacheAfterErrorDuringStateStoreClose() {
         setUpCloseTests();
-        EasyMock.reset(cache);
+        reset(cache);
         cache.flush(CACHE_NAMESPACE);
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: This and the following line 1210 could be removed since they are void returning methods on mocks. Instead verify the invocation of these methods using the function I suggested in one of the comments above.

same for other 2 close tests.

**Review by divijvaidya — CHANGES_REQUESTED** [review]

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 6 above)
Comment (ghost) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Looks like it is not needed, since we are not resetting cachingStore anywhere. Thanks!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 3 above)
Comment (ghost) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:1199]: This will be verified under the method `verifyAndTearDownCloseTests`. Thanks!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 5 above)
Comment (ghost) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Thanks for the suggestion. I have updated the PR

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 4 above)
Comment (ghost) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:94]: Looks like Rule is also not needed.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 8]
```diff
@@ -68,7 +66,7 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 import org.junit.runners.Parameterized.Parameter;
-
+import org.mockito.junit.MockitoJUnitRunner;
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Is this still required? This should be failing checkstyle during build.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 9]
```diff
@@ -1159,58 +1170,40 @@ public void shouldNotThrowInvalidBackwardRangeExceptionWithNegativeFromKey() {
...
-        cache.flush(CACHE_NAMESPACE);
-        cache.close(CACHE_NAMESPACE);
-        EasyMock.replay(cache);
-        EasyMock.reset(underlyingStore);
-        underlyingStore.close();
-        EasyMock.expectLastCall().andThrow(new RuntimeException("Simulating an error on close"));
-        EasyMock.replay(underlyingStore);
-
+        reset(cache);
+        reset(underlyingStore);
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: I don't think we need these reset calls here with Mockito. We required it with easymock because it stacks up the mock stubs but with mockito you can call the same stub as many time as you like.

Same for other close tests.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 10]
```diff
@@ -124,7 +128,8 @@ public static Collection<Object[]> data() {
     public void setUp() {
         baseKeySchema = new TimeFirstWindowKeySchema();
         bytesStore = new RocksDBTimeOrderedWindowSegmentedBytesStore("test", "metrics-scope", 100, SEGMENT_INTERVAL, hasIndex);
-        underlyingStore = new RocksDBTimeOrderedWindowStore(bytesStore, false, WINDOW_SIZE);
+        underlyingStore = spy(new RocksDBTimeOrderedWindowStore(bytesStore,
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: `underlyingStore` is reset only in the close tests (even that is not required, see other comment). In the close tests, the `underlyingStore` member of this class is replaced in the function `setUpCloseTests()` and hence, we don't need to spy this.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 11]
```diff
@@ -144,45 +149,51 @@ public void closeStore() {
...
+        final RocksDBTimeOrderedWindowStore inner = mock(RocksDBTimeOrderedWindowStore.class);
+        when(inner.hasIndex()).thenReturn(hasIndex);
+
+        final TimeOrderedCachingWindowStore outer =
+                spy(new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE,
+                        SEGMENT_INTERVAL));
 
-        EasyMock.reset(inner);
-        EasyMock.expect(inner.name()).andStubReturn("store");
+        reset(inner);
```
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: do we really need a reset of the mock here with Mockito?

**Review by divijvaidya — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Hi @divijvaidya Please review once.

**Review by divijvaidya — APPROVED** [review]
LGTM! @cadonna I have performed an initial review here. Please take a look when you get a chance.

**Comment by ghost** [comment]
Hi  @cadonna,

Just a reminder for the review. Thanks!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 12]
```diff
@@ -104,11 +108,10 @@ public class TimeOrderedCachingPersistentWindowStoreTest {
     private ThreadCache cache;
     private InternalMockProcessorContext context;
     private TimeFirstWindowKeySchema baseKeySchema;
-    private WindowStore<Bytes, byte[]> underlyingStore;
+    private RocksDBTimeOrderedWindowStore underlyingStore;
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Why did you change the type?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 13]
```diff
@@ -124,7 +127,8 @@ public static Collection<Object[]> data() {
     public void setUp() {
         baseKeySchema = new TimeFirstWindowKeySchema();
         bytesStore = new RocksDBTimeOrderedWindowSegmentedBytesStore("test", "metrics-scope", 100, SEGMENT_INTERVAL, hasIndex);
-        underlyingStore = new RocksDBTimeOrderedWindowStore(bytesStore, false, WINDOW_SIZE);
+        underlyingStore = new RocksDBTimeOrderedWindowStore(bytesStore,
+            false, WINDOW_SIZE);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Also this change is not needed, right?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 14]
```diff
@@ -144,45 +148,50 @@ public void closeStore() {
...
-        final RocksDBTimeOrderedWindowStore inner = EasyMock.mock(RocksDBTimeOrderedWindowStore.class);
-        EasyMock.expect(inner.hasIndex()).andReturn(hasIndex);
-        EasyMock.replay(inner);
-        final TimeOrderedCachingWindowStore outer = new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE, SEGMENT_INTERVAL);
+        final RocksDBTimeOrderedWindowStore inner = mock(RocksDBTimeOrderedWindowStore.class);
+        when(inner.hasIndex()).thenReturn(hasIndex);
+
+        final TimeOrderedCachingWindowStore outer =
+                new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE,
+                        SEGMENT_INTERVAL);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Please try to leave everything on the same line if the line does not exceed by too much 120 characters.
```suggestion
        final TimeOrderedCachingWindowStore outer = new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE, SEGMENT_INTERVAL);
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 4 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:94]: In the other migrations to Mockito we specified `@RunWith(MockitoJUnitRunner.StrictStubs.class)`. So if that is not possible here due to restrictions, I think we should use a rule as Divij proposed instead of not using strict stubs at all.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 15]
```diff
@@ -144,45 +148,50 @@ public void closeStore() {
...
 
-        EasyMock.reset(inner);
-        EasyMock.expect(inner.name()).andStubReturn("store");
+        reset(inner);
+        when(inner.name()).thenReturn("store");
         inner.init((org.apache.kafka.streams.processor.ProcessorContext) context, outer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: You need to remove this call since this call is only needed by EasyMock to record the calls to verify. Mockito verifies calls with `verify(inner).init((org.apache.kafka.streams.processor.ProcessorContext) context, outer)`. In this and the next line you are basically calling a method on the mock directly and verifying that you called the method on the mock. That does actually not add anything to the test. In a test you should verify that a call to a method of the class under test calls the method on the mock, not that the test calls directly the method on the mock.
So you can remove both line. 
After that you need also to remove the `times(2)` in `verify(inner, times(2)).init((org.apache.kafka.streams.processor.ProcessorContext) context, outer);` since that is only needed, because you call the method on the mock directly.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 16]
```diff
@@ -144,45 +148,50 @@ public void closeStore() {
...
+        final RocksDBTimeOrderedWindowStore inner = mock(RocksDBTimeOrderedWindowStore.class);
+        when(inner.hasIndex()).thenReturn(hasIndex);
+
+        final TimeOrderedCachingWindowStore outer =
+                new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE,
+                        SEGMENT_INTERVAL);
 
-        EasyMock.reset(inner);
-        EasyMock.expect(inner.name()).andStubReturn("store");
+        reset(inner);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: This is not needed with Mockito.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java (hunk 15 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: So this test should look like this:
```java
    @Test
    public void shouldDelegateDeprecatedInit() {
        final RocksDBTimeOrderedWindowStore inner = mock(RocksDBTimeOrderedWindowStore.class);
        when(inner.hasIndex()).thenReturn(hasIndex);

        final TimeOrderedCachingWindowStore outer = new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE, SEGMENT_INTERVAL);
        outer.init((org.apache.kafka.streams.processor.ProcessorContext) context, outer);

        verify(inner).init((org.apache.kafka.streams.processor.ProcessorContext) context, outer);
    }
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 17]
```diff
@@ -144,45 +148,50 @@ public void closeStore() {
...
+        verify(inner).init((org.apache.kafka.streams.processor.ProcessorContext) context, outer);
+
         outer.init((org.apache.kafka.streams.processor.ProcessorContext) context, outer);
-        EasyMock.verify(inner);
+        verify(inner, times(2))
+                .init((org.apache.kafka.streams.processor.ProcessorContext) context, outer);
     }
 
     @Test
     public void shouldDelegateInit() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java:161]: See my comments above.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 18]
```diff
@@ -1235,4 +1220,15 @@ private int addItemsToCache() {
...
 
+    private void verifyAndTearDownCloseTests() {
+        verify(underlyingStore).close();
+        verify(cache).flush(CACHE_NAMESPACE);
+        verify(cache).close(CACHE_NAMESPACE);
+
+        // resets the mocks created in #setUpCloseTests(). It is necessary to
+        // ensure that @After works correctly.
+        reset(cache);
+        reset(underlyingStore);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: You do not need this if you use add a `doNothing()` where you specify the exceptions to be thrown. For example, 
```java
doThrow(new RuntimeException("Simulating an error on flush2")).when(cache).flush(CACHE_NAMESPACE);
```
becomes

```java
doThrow(new RuntimeException("Simulating an error on flush2")).doNothing().when(cache).flush(CACHE_NAMESPACE);
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 19]
```diff
@@ -144,45 +148,50 @@ public void closeStore() {
...
     public void shouldDelegateInit() {
-        final RocksDBTimeOrderedWindowStore inner = EasyMock.mock(RocksDBTimeOrderedWindowStore.class);
-        EasyMock.expect(inner.hasIndex()).andReturn(hasIndex);
-        EasyMock.replay(inner);
-        final TimeOrderedCachingWindowStore outer = new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE, SEGMENT_INTERVAL);
+        final RocksDBTimeOrderedWindowStore inner = mock(RocksDBTimeOrderedWindowStore.class);
+        when(inner.hasIndex()).thenReturn(hasIndex);
+
+        final TimeOrderedCachingWindowStore outer =
+                spy(new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE,
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Why do you need a spy here?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 20]
```diff
@@ -53,11 +53,9 @@
 import org.apache.kafka.streams.state.internals.PrefixedWindowKeySchemas.TimeFirstWindowKeySchema;
 import org.apache.kafka.test.InternalMockProcessorContext;
 import org.apache.kafka.test.TestUtils;
-import org.easymock.EasyMock;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
-
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: Why did you remove those empty lines? Please put them back.

**Review by cadonna — COMMENTED** [review]
@shekhar-rajak Thanks for the PR and thank you for your patience!

Here is my feedback!

**Comment by cadonna** [comment]
@shekhar-rajak Do you have any updates for this PR?

**Comment by clolov** [comment]
@shekhar-rajak, I would quite like to complete the migration for streams, so I will wait for a response for another week. Afterwards I will open a separate pull request based on this one in order to get it merged.

**Comment by ghost** [comment]
Ack. Let me update the PR

**Comment by guozhangwang** [comment]
@shekhar-rajak please let me know after you've updated the PR incorporating previous comments.

**Comment by ghost** [comment]
@guozhangwang  Please review and merge. Thanks!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 21]
```diff
@@ -144,45 +148,35 @@ public void closeStore() {
...
-        EasyMock.reset(inner);
-        EasyMock.expect(inner.name()).andStubReturn("store");
-        inner.init((StateStoreContext) context, outer);
-        EasyMock.expectLastCall();
-        EasyMock.replay(inner);
+        final TimeOrderedCachingWindowStore outer = new TimeOrderedCachingWindowStore(inner, WINDOW_SIZE, SEGMENT_INTERVAL);
         outer.init((StateStoreContext) context, outer);
-        EasyMock.verify(inner);
+        verify(inner, times(1)).init((StateStoreContext) context, outer);
+
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: nit:
```suggestion
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java [hunk 22]
```diff
@@ -1159,58 +1153,35 @@ public void shouldNotThrowInvalidBackwardRangeExceptionWithNegativeFromKey() {
...
-        EasyMock.expect(underlyingStore.name()).andStubReturn("store-name");
-        EasyMock.expect(underlyingStore.isOpen()).andStubReturn(true);
-        EasyMock.replay(underlyingStore);
-        cachingStore = new TimeOrderedCachingWindowStore(underlyingStore, WINDOW_SIZE, SEGMENT_INTERVAL);
-        cache = EasyMock.createNiceMock(ThreadCache.class);
+        underlyingStore = mock(RocksDBTimeOrderedWindowStore.class);
+        when(underlyingStore.name()).thenReturn("store-name");
+        when(underlyingStore.isOpen()).thenReturn(true);
+        cachingStore = new TimeOrderedCachingWindowStore(underlyingStore,
+            WINDOW_SIZE, SEGMENT_INTERVAL);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedCachingPersistentWindowStoreTest.java]: nit:
```suggestion
        cachingStore = new TimeOrderedCachingWindowStore(underlyingStore, WINDOW_SIZE, SEGMENT_INTERVAL);
```

**Review by cadonna — APPROVED** [review]
Thanks for the updates, @shekhar-rajak !

LGTM!

I have some nits that I will commit myself.

**Comment by cadonna** [comment]
Once the build are acceptable, I am going to merge this PR.

**Comment by cadonna** [comment]
Build failures are unrelated.
